ArcGISRuntimeSDKQt_CppSamples/Maps/MobileMap_SearchAndRoute/MobileMap_SearchAndRoute.cpp

```diff
@@ -0,0 +1,366 @@
+#include "RouteParameters.h"
+#include "SimpleLineSymbol.h"
+#include "MobileMapPackage.h"
+#include "PictureMarkerSymbol.h"
+#include "ReverseGeocodeParameters.h"
+
+#include <QDir>
+#include <QQmlProperty>
+#include <QFileInfoList>
+#include <QFile>
+
+using namespace Esri::ArcGISRuntime;
+
+MobileMap_SearchAndRoute::MobileMap_SearchAndRoute(QQuickItem* parent):
+    QQuickItem(parent),
+    m_map(nullptr),
+    m_mapView(nullptr),
+    m_calloutData(nullptr),
+    m_bluePinSymbol(nullptr),
+    m_currentRouteTask(nullptr),
+    m_currentLocatorTask(nullptr),
+    m_stopsGraphicsOverlay(nullptr),
+    m_routeGraphicsOverlay(nullptr),
+    m_canRoute(false),
+    m_canClear(false),
+    m_isGeocodeInProgress(false)
+{
+}
+
+MobileMap_SearchAndRoute::~MobileMap_SearchAndRoute()
+{
+}
+
+void MobileMap_SearchAndRoute::componentComplete()
+{
+    QQuickItem::componentComplete();
+
+    // find QML MapView component
+    m_mapView = findChild<MapQuickView*>("mapView");
+    m_mapView->setWrapAroundMode(WrapAroundMode::Disabled);
+
+    m_fileInfoList = QDir(QDir::homePath() + "/ArcGIS/Runtime/Data/mmpk/").entryInfoList();
+
+    // initialize Callout
+    m_mapView->calloutData()->setTitle("Address");
+    m_calloutData = m_mapView->calloutData();
+    emit calloutDataChanged();
+
+    // set reverse geocoding parameters
+    m_reverseGeocodeParameters.setMaxResults(1);
+    QStringList resultAttributeNames;
+    resultAttributeNames << "Address" << "Neighborhood" << "City" << "Region" << "Street";
+    m_reverseGeocodeParameters.setResultAttributeNames(resultAttributeNames);
+
+    // identify and create MobileMapPackages using mmpk files in datapath
+    createMobileMapPackages(0);
+
+    // create graphics overlays to visually display geocoding and routing results
+    m_stopsGraphicsOverlay = new GraphicsOverlay(this);
+    m_routeGraphicsOverlay = new GraphicsOverlay(this);
+    m_routeGraphicsOverlay->setRenderer(new SimpleRenderer(new SimpleLineSymbol(SimpleLineSymbolStyle::Solid, QColor("#2196F3"), 4, this), this));
+
+    m_mapView->graphicsOverlays()->append(m_stopsGraphicsOverlay);
+    m_mapView->graphicsOverlays()->append(m_routeGraphicsOverlay);
+
+    // create a pin symbol
+    m_bluePinSymbol = new PictureMarkerSymbol(QUrl("qrc:/Samples/Maps/MobileMap_SearchAndRoute/bluePinSymbol.png"), this);
+    m_bluePinSymbol->setHeight(36);
+    m_bluePinSymbol->setWidth(36);
+    m_bluePinSymbol->setOffsetY(m_bluePinSymbol->height() / 2);
+
+    connectSignals();
+}
+
+void MobileMap_SearchAndRoute::createMobileMapPackages(int index)
+{
+    if (index < m_fileInfoList.length())
+    {
+        // check if file is a .mmpk file
+        if (m_fileInfoList[index].completeSuffix() == "mmpk")
+        {
+            // create a new MobileMapPackage
+            MobileMapPackage* mobileMapPackage = new MobileMapPackage(m_fileInfoList[index].absoluteFilePath());
+            // load the new MMPK
+            mobileMapPackage->load();
```

Comment:
It's better if this call is made after the connection is setup.  So, you can move it after the connection logic.  There might be a chance that the call completes before it hits the connection logic if it comes later.  However, if it setup before the call, then it is safer.